pkg/download/consistent_hashing.go

```diff
@@ -154,80 +169,77 @@ func (m *ConsistentHashingMode) Fetch(ctx context.Context, urlString string) (io
 		chunksPerSlice = append([]int64{0}, EqualSplit(int64(concurrency), totalSlices-1)...)
 	}
 
+	readersCh := make(chan io.Reader, m.maxConcurrency()+1)
+	readersCh <- br
+
 	logger.Debug().Str("url", urlString).
 		Int64("size", fileSize).
 		Int("concurrency", m.maxConcurrency()).
 		Ints64("chunks_per_slice", chunksPerSlice).
 		Msg("Downloading")
 
-	for slice, numChunks := range chunksPerSlice {
-		if numChunks == 0 {
-			// this happens if we've already downloaded the whole first slice
-			continue
-		}
-		startFrom := m.SliceSize * int64(slice)
-		sliceSize := m.SliceSize
-		if slice == 0 {
-			startFrom = firstChunkResp.ContentLength
-			sliceSize = sliceSize - firstChunkResp.ContentLength
-		}
-		if slice == int(totalSlices)-1 {
-			sliceSize = (fileSize-1)%m.SliceSize + 1
-		}
-		if sliceSize/numChunks < m.minChunkSize() {
-			// reset numChunks to respect minChunkSize
-			numChunks = sliceSize / m.minChunkSize()
-			// although we must always have at least one chunk
+	m.queue.submit(func() {
+		defer close(readersCh)
```

Comment:
Same as the other comment -- this is a lot of code in the closure. I'd like it to be isolated to external function eventually.